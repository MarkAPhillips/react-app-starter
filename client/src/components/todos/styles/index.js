import styled from 'styled-components';
import { colours } from '../../../assets/styles/variables';

export const ContainerPanel = styled.div`
  &:hover {
    cursor: ${props => (props.isAction ? 'pointer' : 'auto')};
  }
`;

export const TodoContainerPanel = styled.section`
  border-radius: 7px;
  background-color: ${colours.lightgrey};
  padding: 1em;
`;

export const TodoListItemPanel = styled.div`
  display: flex;
  padding: 0.5em;
  align-items: center;
  width: 90%;
  justify-content: space-between;
`;

export const TodoListPanel = styled.div`
  padding: 0.5em;
  background-color: ${colours.white};
  margin: 12px 0;
  border-radius: 3px;
  > div {
    border-bottom: 1px solid ${colours.lightgrey};
    margin-bottom: 8px;
    &:last-child {
      border-bottom: none;
      margin-bottom: 0;
    } 
    &:hover {
      opacity: 0.7;
    }
  }
`;

export const TodoFilterPanel = styled.div``;

export const InputPanel = styled.div`
  width: 400px;
`;
