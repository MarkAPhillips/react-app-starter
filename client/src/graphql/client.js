import ApolloClient from 'apollo-boost';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { API } from '../constants';

const cache = new InMemoryCache({
  dataIdFromObject: o => o.id,
});

export const client = new ApolloClient({
  uri: API.ROOT_URL,
  cache,
});
