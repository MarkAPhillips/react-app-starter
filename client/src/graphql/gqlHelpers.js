import gql from 'graphql-tag';
import { print } from 'graphql/language/printer';

export const convertQueryToString = query => gql`${print(query)}`;
