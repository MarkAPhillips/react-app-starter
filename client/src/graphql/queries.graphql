query getAll {
  todos {
    id
    item
    createdDate
    completed
  }
}

