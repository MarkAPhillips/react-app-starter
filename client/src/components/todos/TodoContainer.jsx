import React, { useState } from 'react';
import { TodoContainerPanel } from './styles';
import { Subheader } from '../../assets/styles/components';
import { TodoForm, TodoList } from './';


export const TodoContainer = () => {
  const { inputValue, setInputValue } = useState('');

  const handleChange = (evt) => {
    const { value } = evt.target;
    console.log('handleChange', inputValue);
    setInputValue(value);
  };

  const handleSubmit = (evt) => {
    evt.preventDefault();
    console.log('handleSubmit', inputValue);
    setInputValue('');
  };

  const handleStatusChange = (evt) => {
    const { checked, id } = evt.target;
    console.log('handleStatusChage', checked, id);
  };

  return (
    <TodoContainerPanel>
      <Subheader>Inbox</Subheader>
      <TodoList
        onStatusChange={handleStatusChange}
      />
      <TodoForm
        onSubmit={handleSubmit}
        onChange={handleChange}
        inputValue={inputValue}
      />
    </TodoContainerPanel>
  );
};
