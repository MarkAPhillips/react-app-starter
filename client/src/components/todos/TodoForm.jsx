import React, { useRef } from 'react';
import { Mutation } from 'react-apollo';
import { Input, PrimaryButton } from '../../assets/styles/components';
import { InputPanel } from './styles';
import { create } from '../../graphql/mutations.graphql';
import { getAll } from '../../graphql/queries.graphql';
import { convertQueryToString } from '../../graphql/gqlHelpers';

const query = convertQueryToString(getAll);

export const TodoForm = () => {
  const input = useRef('');
  return (
    <Mutation
      mutation={create}
      update={(cache, { data: { createTodo } }) => {
        const { todos } = cache.readQuery({ query });
        cache.writeQuery({
          query,
          data: { todos: todos.concat([createTodo]) },
        });
      }}
    >
      {createMutation => (
        <div>
          <form onSubmit={(e) => {
            e.preventDefault();
            createMutation({ variables: { text: input.current.value } });
            input.current.value = '';
          }}
          >
            <InputPanel>
              <Input
                type="text"
                id="input-add-todo"
                ref={input}
                placeholder="e.g. Complete Typescript course"
                required
              />
            </InputPanel>
            <PrimaryButton type="submit">Add</PrimaryButton>
          </form>
        </div>
      )}
    </Mutation>
  );
};
