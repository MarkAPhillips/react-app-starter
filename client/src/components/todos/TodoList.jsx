import React, { Fragment } from 'react';
import _ from 'lodash';
import { Query } from 'react-apollo';
import { getAll } from '../../graphql/queries.graphql';
import { TodoListItem } from '.';
import { Subheader } from '../../assets/styles/components';
import { TodoListPanel } from './styles';

export const TodoList = () => (
  <Query query={getAll}>
    {({ loading, error, data }) => {
      if (loading) return 'Loading...';
      if (error) return `Error! ${error.message}`;
      if (_.isEmpty(data.todos)) return null;
      return (
        <Fragment>
          <Subheader>Inbox</Subheader>
          <TodoListPanel>
            {data.todos.map(todo => (
              <TodoListItem
                key={todo.id}
                todo={todo}
              />
            ))}
          </TodoListPanel>
        </Fragment>
      );
    }}
  </Query>
);
