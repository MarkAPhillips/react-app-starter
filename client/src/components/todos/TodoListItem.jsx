import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { remove, update } from '../../graphql/mutations.graphql';
import { getAll } from '../../graphql/queries.graphql';
import { TodoListItemPanel, ContainerPanel } from './styles';
import { CheckBox } from '../shared';
import { convertQueryToString } from '../../graphql/gqlHelpers';

const defaultProps = {
  todo: {},
};

const query = convertQueryToString(getAll);

const propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.string,
    item: PropTypes.string,
    completed: PropTypes.bool,
  }),
};

export const TodoListItem = ({ todo }) => {
  const [isHovering, setHoveringState] = useState(false);

  const handleMouseHover = (currentHovering) => {
    if (isHovering !== currentHovering) {
      setHoveringState(() => currentHovering);
    }
  };

  const { id, completed, item } = todo;

  return (
    <TodoListItemPanel
      onMouseEnter={() => handleMouseHover(true)}
      onMouseLeave={() => handleMouseHover(false)}
    >
      <ContainerPanel>
        <Mutation mutation={update}>
          { updateMutation => (
            <CheckBox
              id={id}
              completed={completed}
              item={item}
              onChange={(e) => {
                const { checked } = e.target;
                updateMutation({ variables: { id, completed: checked } });
              }}
            />)
          }
        </Mutation>
      </ContainerPanel>
      {isHovering &&
        <Mutation
          mutation={remove}
          variables={{ id }}
          update={(cache) => {
            const { todos } = cache.readQuery({ query });
            cache.writeQuery({
              query,
              data: { todos: todos.filter(e => e.id !== id) },
            });
          }}
        >
          { deleteMutation => (
            <ContainerPanel
              onClick={deleteMutation}
              isAction
            > <FontAwesomeIcon icon="trash-alt" />
            </ContainerPanel>) }
        </Mutation>}
    </TodoListItemPanel>
  );
};

TodoListItem.propTypes = propTypes;
TodoListItem.defaultProps = defaultProps;
