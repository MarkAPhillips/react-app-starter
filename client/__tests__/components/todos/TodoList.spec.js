import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import { getAll } from '../../../src/graphql/queries.graphql';
import { componentSnapshotShouldMatch } from '../../testUtils';
import { TodoList } from '../../../src/components/todos';

describe('TodoList specs', () => {
  const todos = [{
    id: '71cda7b6-749b-47ec-937c-dca827dcd218',
    completed: false,
    item: 'Todo item',
  }];

  const mocks = [
    {
      request: {
        query: getAll,
      },
      result: {
        data: {
          todos,
        },
      },
    },
  ];

  it('should render the component ', () => {
    const component =
      (
        <MockedProvider mocks={mocks} addTypename={false}>
          <TodoList />
        </MockedProvider>);

    componentSnapshotShouldMatch((component));
  });
});
