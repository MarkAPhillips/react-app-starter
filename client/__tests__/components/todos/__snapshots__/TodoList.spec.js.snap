// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList specs should render the component  1`] = `"Loading..."`;
